File: app_py/routers/permissions.py
from fastapi import APIRouter, Depends, Request, Form, HTTPException
from fastapi.responses import RedirectResponse, HTMLResponse, JSONResponse
from sqlalchemy.orm import Session

from app_py.database import get_db
from app_py.dependencies import require_permission, get_menu_form_groups
from sqlalchemy.orm import joinedload
from app_py.models import CmsUser, Form, UserGroup, GroupPermission
from app_py.templating import templates, get_context
from app_py.utils import url_for

router = APIRouter()


@router.get("/{group_id}", response_class=HTMLResponse)
async def edit(
    request: Request,
    group_id: int,
    current_user: CmsUser = Depends(require_permission("users", "read")),
    db: Session = Depends(get_db),
):
    group = db.query(UserGroup).options(joinedload(UserGroup.permissions)).filter(UserGroup.id == group_id).first()
    if not group:
        raise HTTPException(404)
    forms = db.query(Form).order_by(Form.sort_order).all()
    tables = {"_forms": "表单管理", "_users": "用户管理"}
    for f in forms:
        tables[f.table_name] = f.name
    perms = {p.table_name: p for p in group.permissions}
    menu = get_menu_form_groups(db, current_user)
    return templates.TemplateResponse(
        "permissions/edit.html",
        get_context(request, cms_user=current_user, menu_form_groups=menu, group=group, tables=tables, perms=perms),
    )


def _parse_permissions_form(form_data: dict) -> dict:
    result = {}
    for key, val in form_data.items():
        if key.startswith("permissions[") and "][" in key and key.endswith("]"):
            inner = key[len("permissions["):-1]
            parts = inner.split("][")
            if len(parts) == 2:
                table, action = parts
                if table not in result:
                    result[table] = {}
                result[table][action] = val not in ("", "0", "false", False)
    return result


@router.put("/{group_id}")
async def update(
    request: Request,
    group_id: int,
    current_user: CmsUser = Depends(require_permission("users", "update")),
    db: Session = Depends(get_db),
):
    group = db.query(UserGroup).filter(UserGroup.id == group_id).first()
    if not group:
        raise HTTPException(404)
    form_data = await request.form()
    submitted = _parse_permissions_form(dict(form_data))
    forms = db.query(Form).order_by(Form.sort_order).all()
    all_tables = {"_forms": 1, "_users": 1}
    for f in forms:
        all_tables[f.table_name] = 1
    for table_name in all_tables:
        actions = submitted.get(table_name, {})
        perm = db.query(GroupPermission).filter(
            GroupPermission.user_group_id == group_id,
            GroupPermission.table_name == table_name,
        ).first()
        if not perm:
            perm = GroupPermission(user_group_id=group_id, table_name=table_name)
            db.add(perm)
        perm.can_create = bool(actions.get("create"))
        perm.can_read = bool(actions.get("read"))
        perm.can_update = bool(actions.get("update"))
        perm.can_delete = bool(actions.get("delete"))
    db.commit()
    return JSONResponse({"code": 0, "msg": "权限保存成功"})

File: internal/handlers/permissions.go
package handlers

import (
	"laycms/internal/models"
	"laycms/internal/tmpl"
	"strings"

	"github.com/gin-gonic/gin"
)

type tablePerm struct {
	Name  string
	Label string
}

func permissionsEditPage(c *gin.Context) {
	u := CurrentUser(c)
	id := atoi(c.Param("group_id"), 0)
	var g models.UserGroup
	if err := gormDB().Preload("Permissions").First(&g, id).Error; err != nil {
		c.Status(404)
		return
	}
	var forms []models.Form
	gormDB().Order("sort_order ASC").Find(&forms)
	tables := []tablePerm{
		{"_forms", "表单管理"},
		{"_users", "用户管理"},
	}
	for _, f := range forms {
		tables = append(tables, tablePerm{f.DataTable, f.Name})
	}
	permMap := map[string]*models.GroupPermission{}
	for i := range g.Permissions {
		p := &g.Permissions[i]
		permMap[p.ResourceTable] = p
	}
	page := tmpl.BuildPage(c, gormDB(), u, "权限配置 - "+g.Name, "")
	type frag struct {
		*tmpl.Page
		Group   models.UserGroup
		Tables  []tablePerm
		Perms   map[string]*models.GroupPermission
	}
	_ = tmpl.RenderFragmentWith(c, "permissions_edit", &frag{page, g, tables, permMap}, page)
}

func permissionsUpdate(c *gin.Context) {
	id := atoi(c.Param("group_id"), 0)
	var g models.UserGroup
	if err := gormDB().First(&g, id).Error; err != nil {
		c.Status(404)
		return
	}
	_ = c.Request.ParseForm()
	submitted := parsePermissionsForm(c.Request.PostForm)
	var forms []models.Form
	gormDB().Order("sort_order ASC").Find(&forms)
	allTables := map[string]struct{}{"_forms": {}, "_users": {}}
	for _, f := range forms {
		allTables[f.DataTable] = struct{}{}
	}
	for tn := range allTables {
		actions := submitted[tn]
		var perm models.GroupPermission
		q := gormDB().Where("user_group_id = ? AND table_name = ?", id, tn)
		if err := q.First(&perm).Error; err != nil {
			perm = models.GroupPermission{
				UserGroupID: int64(id), ResourceTable: tn,
				CanCreate: actions["create"],
				CanRead:   actions["read"],
				CanUpdate: actions["update"],
				CanDelete: actions["delete"],
			}
			gormDB().Create(&perm)
			continue
		}
		perm.CanCreate = actions["create"]
		perm.CanRead = actions["read"]
		perm.CanUpdate = actions["update"]
		perm.CanDelete = actions["delete"]
		gormDB().Save(&perm)
	}
	c.JSON(200, gin.H{"code": 0, "msg": "权限保存成功"})
}

func parsePermissionsForm(vs map[string][]string) map[string]map[string]bool {
	out := map[string]map[string]bool{}
	prefix := "permissions["
	for key, vals := range vs {
		if !strings.HasPrefix(key, prefix) || !strings.HasSuffix(key, "]") {
			continue
		}
		inner := key[len(prefix) : len(key)-1]
		parts := strings.Split(inner, "][")
		if len(parts) != 2 {
			continue
		}
		table, action := parts[0], parts[1]
		if out[table] == nil {
			out[table] = map[string]bool{}
		}
		val := ""
		if len(vals) > 0 {
			val = vals[len(vals)-1]
		}
		out[table][action] = truthy(val)
	}
	return out
}

func truthy(v string) bool {
	return v != "" && v != "0" && v != "false"
}
